### ISSUE TITLE
CovPassCheck shows warning "Check date and time"

### ISSUE THREAD

**Issue by jkrwdf** [Digitaler-Impfnachweis/covpass-android#97]
## Avoid duplicates
* [X] Bug is not mentioned in the [FAQ](https://digitaler-impfnachweis-app.de/faq/)
* [X] Bug is not already reported in another issue

## Technical details

- Device name: Google Pixel 3 XL
- OS version: Android 12
- App version: 1.12.0

## Describe the bug

Since the update just installed, CovPassCheck shows a warning that I should adjust my device time. 
See screenshot.

![image](https://user-images.githubusercontent.com/5329491/140030102-534f72f4-8c91-4465-bcba-5c95fb95e2b4.png)

In the screenshot it says "It should be 2 Nov 2021 09:47", although the current device date and time (3 Nov, 09:35) is correct.

## Steps to reproduce the issue

Start the app. 

## Expected behaviour

Clear.

**Comment by Ein-Tim** [comment]
I can't reproduce this after updating to 1.12. No warning is shown.

Huawei P Smart 2019 with Android 10.

**Comment by jkrwdf** [comment]
I have a second device, Google Pixel 4a (5G), also on Android 12. 
There I can't reproduce the issue either.

I rebooted my Google Pixel 3 XL, and afterwards the warning is gone there also.

**Comment by Jo-Achim** [comment]
Temporary same problem here after restarting my smartphone (Samsung Galaxy Note 10, SM-N970F / DS /// Android 11).
Now (Smartphone): 05.11.2021 10:15
CovPassCheck 1.12.0:
- Check Date an Time
It should be: 03.11.2021 23:22 (this may be the update time from CovPassCheck 1.11.1 to 1.12.0.)

But after some minutes and CovPassCheck restart: No more "Check Date an Time" messages.

**Comment by Jo-Achim** [comment]
CovPassCheck 1.12.0:
Currently (11/13/2021 02:21 pm) the same "Check date and time" error.
"Soll: 05.11.2021 13:41".

_Notes:_ 
- "Update check rules", "Update" = "List of certificate issuers: 13.11.2021 14:23" and "Check rules: 13.11.2021 14:23" did not fix the "Check date and time" error.
- Android 11 (UI 3.1) was updated today at 10:51 a.m. (security update from 10/01/2021 to version from 11/01/2021. 
- An additional restart (after installing the Samsung update) fixed the "Check date and time" error.

**Comment by pucherd** [comment]
Hello @Jo-Achim,
Thank you for your latest update. Can you please provide us more inputs to your first comment?
The issue appeared after you updated CovPassCheck from 1.11.1 to 1.12.0? Have you switched to flight mode after the update?

**Comment by Jo-Achim** [comment]
Hello @pucherd, 

regarding my first comment (11/05/2021):
- yes, the error occurred immediately after updating CovPassCheck from v1.11.1 to 1.12.0 (Android).
- no, I am very sure that I did not switch to flight mode in this situation (and that flight mode was also switched off).


On my second comment (11/13/2021):
Here, too, the flight mode was certainly switched off.

(Info: I use the flight mode very rarely and only in special environments (e.g. doctor, etc.) - but then I always have Bluetooth switched on manually with regard to CWA.)

**Comment by Jo-Achim** [comment]
Hello @pucherd,

I have a additional querstion: 
Is it right, that CovPassCheck v1.12.0 uses a time-server? **1.de.pool.ntp.org/123**?
Would it be conceivable that the sporadic problem "check date and time" could have something to do with small time differences between ntp (server) and the time of the cell phone?

**Comment by pucherd** [comment]
Hi @Jo-Achim,
The CovPassCheck is using a time-server **1.de.pool.ntp.org**. A small time difference between the ntp server and the cell phone should not be an issue because the defined maximum offset is 2 hours.

**Comment by jkrwdf** [comment]
At least for my occurrence of this issue (once in one of my phones), I can inform that at the very time when the issue was present, the time difference between my phone and the NTP time was below 200 ms (checked with https://uhr.ptb.de). Of course I checked my phone date and time first before opening this issue. 

But this (for me new) information opens room for a bug in CovPassCheck where and outdated NTP time information is not updated properly under certain circumstances (update of CovPassCheck, update of phone software) and instead of getting the time fresh, the software believes the time is still fresh, continues, and then in later code runs into this time-difference check against the phone time. Note that in all cases, the "should be..." time was in all cases far back in the past.

**Comment by Jo-Achim** [comment]
> 
> @jkrwdf wrote:
> But this (for me new) information opens room for a bug in CovPassCheck where an outdated NTP time information is not updated properly under certain circumstances (update of CovPassCheck, update of phone software) and instead of getting the time fresh, the software believes the time is still fresh, continues, and then in later code runs into this time-difference check against the phone time. Note that in all cases, the "should be..." time was far back in the past.

So for me the explanation of @jkrwdf would be logical and would confirm my basic assumption.

**Comment by timokoenig** [comment]
The NTP is being used in CovPass and CovPassCheck to indicate if the user changed their time on accident. Especially in CovPassCheck this would cause false check results therefore this little message should remind the user to check the local time.

It is not unusual that different NTP server are sometimes out of sync but in this case where the difference is roughly one day it sounds more like a cache issue

**Comment by Ein-Tim** [comment]
FYI: CWA also uses a time server. If your device time is off more than 2h, no risk checks are carried out.

**Comment by Jo-Achim** [comment]
> 
> @timokoenig wrote:
> The NTP is being used in CovPass and CovPassCheck ...

... then as a precaution I would like to add a cross-reference to a possibly similar problem: https://github.com/Digitaler-Impfnachweis/covpass-android/issues/98#issue-1044602167.

**Comment by timokoenig** [comment]
I would guess we have different causes for those two issues but that needs to be checked

FYI @alexcimander

**Comment by pucherd** [comment]
#98 is not related to this issue.
I think the NTP issue could be related to this already reported bug from TrueTime library:
https://github.com/instacart/truetime-android/issues/105

**Comment by smrqdt** [comment]
I have the same problem on an OnePlus 7t on Android 11 with CovPassCheck 1.12.0.

Right now (2021-11-17 21:49) it shows after restarting the app
> It should be: 13 Nov 2021 06:34

Seems to be a static time difference of -5 days, 8 hours, 44 minutes or 400506 ± 60 seconds, so I don’t know in which unit or number system this makes sense…

By the way, if you’re using pool.ntp.org timeservers in an application by default you **have** to apply for vendor specific hostnames:
[pool.ntp.org: The NTP Pool for vendors](https://www.pool.ntp.org/vendors.html)

**Comment by jkrwdf** [comment]
I do not want to destroy the example for the experts, but before you plan to restart your phone anyway, maybe you can first try to delete the applications cache data (Settings / Apps / CovPassCheck / Speicher und Cache / Cache leeren" (too lazy to switch to EN right now) to test whether this ad-hoc solves the issue. @timokoenig supposes a cache issue in https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issuecomment-971493037.

**Comment by jkrwdf** [comment]
Yesterday I again had the issue (it was December 1st, 17:11, and the app claimed it is 30.11.2021 22:03). 

The phone was quite freshly restarted (somewhen in the afternoon). I cannot remember anything specific which happened at 22:03 the day before, certainly neither app or OS update). 

**Force-close** and restart the app did **not** help.
**Deleting the app's cache data** via Android settings did **not** help.
**Deleting the app's entire storage** and cache of course **did** help.

**Comment by Ein-Tim** [comment]
Confirm same issue on my Huawei P Smart 2019. It shows "It should be: 28.11.2021 14:47"

Any fix for this in sight?

**Comment by SandraEickel** [comment]
Xiaomi Mi 10T Lite 5G with MIUI 12.5.6 (Android 11, Patch level 2021-09-01) here, fresh CovPass-Check 1.13.3 still shows this bogus warning to me (I should turn back time to end of October). Nearly six weeks now since first report of this issue ...

**Comment by SandraEickel** [comment]
> I think the NTP issue could be related to this already reported bug from TrueTime library: [instacart/truetime-android#105](https://github.com/instacart/truetime-android/issues/105)

There have been several other bugs reported for different aspects of this issue with the TrueTime library during the last three years, but there has been almost no development of it at all (even if there is a new version 3.5 since October).

Summary of what I've picked up by reading through those issues:
The intention of this library is to detect manual time modifications by the end user without requiring a permanent internet connection to a network time server.
The library wants to be notified when a reboot is finished, but that fails now on many systems since Android "Nougat" and depends somehow on the device vendor. So the library gets the correct time only once when it is installed and used for the first time. That time is stored. Further calculations are based on the stored value plus up-time of the device.
The next reboot changes everything, If the library is notified, it fetches a new time via NTP and updates the cached value, so the new calculations work. If not, the result is this bogus message.
There is an API to delete the cached value, but that doesn't seem to work reliable, either. This might be the easiest way to get to a reliable workaround - as long as the device is not offline for a longer time, but then rules (and revocations) are outdated, anyway (I'm using CovPass-Check for verifying and CoronaWarnApp for every other purpose).
There are 169 forks of truetime-android, perhaps at least one of them has been fixed in one or the other way.

**Comment by nnc-tlb** [comment]
I have this issue on my device xiaomi redmi note 7 (PixelExperience_Plus_lavender-11.0-20210915-1722-OFFICIAL) since nearly 2 weeks

covpasscheck is version 1.13.3

'Ist'-Zeit = 13.12.2021 02:00
'Soll'-Zeit = 24.11.2021 03:30

**Comment by wkornewald** [comment]
@nnc-tlb and anyone with Xiami phones: Could you please check if going to the app's security/permissions and turning on the Autostart toggle works? See https://i.stack.imgur.com/pXzyt.png and https://stackoverflow.com/questions/35718468/boot-broadcastreceiver-does-not-work-on-xiaomi-devices

@SandraEickel thanks for the analysis. I think we'll have to force a re-sync of the NTP time on app start if there's an Internet connection and otherwise maybe somehow detect a reboot.

**Comment by nnc-tlb** [comment]
Because I did a de-install / re-install I cannot check the old settings.
After re-install and first run of the covpasscheck-app w/o any permissions the "date/time warning box" was gone.

On first QR-code scan I gave the covpasscheck-app only permission to the camera (checked that again in settings, still correct).

On all subsequent runs (till now) still all good.

**Comment by wkornewald** [comment]
@nnc-tlb You could probably repeat the bug by rebooting your device. Maybe it only gets triggered by (quick) reboot instead of turning off + on.

At least on HTC the issue probably only happens with reboot (there we can observe a special reboot event to fix the issue, but Xiaomi is different again).

**Comment by SandraEickel** [comment]
@wkornewald
Yuck. Throw in some random ... I did NOT enable Autostart, did a Restart and a Power Off/On cycle.
In both reboot cases, CovPass-Check didn't show me the bogus warning that I had before.
So I can't check if enabling Autostart would be a reliable fix :-(.

The only thing I could try to get the bug back would be another reboot and then perhaps waiting a day or so before running CovPass-Check for the first time - the usage of it just a short time after reboot could have triggered another check for a required update in the TrueTime library.

Sidenote: Rules were not updated during these tests, as that update will only be done if more than 24h have passed since last update (manual triggering that would be a nice enhancement).

**Comment by Jo-Achim** [comment]
If the problem lies outside of the direct access (e.g. external (NTP-) library), there maybe the inelegant wooden hammer method as a workaround: Whenever the app wants to display the error and there is an internet connection ... carry out a time adjustment (or work with the current smartphone time).

**Comment by SandraEickel** [comment]
I have the bug back (reboot last night, first run of CovPass-Check this afternoon).

So I've turned on Autostart for CovPass-Check and am rebooting now.
Tomorrow, I'm going to see if CovPass-Check complains about time or if it's happy with it.
[...]
Preliminary result: Enabling Autostart for CovPass-Check seems to fix the problem my Mi brand family smart phone.

@jkrwdf You are predestined to do some confirming tests :-).
a) Run CovPass(-Check) within ten minutes or up to slightly more than one hour after power on (with Autostart disabled, if your phone Operating System has that setting).
b) Turn Autostart for the application on before power off and check the next day more than two hours after reboot if CovPass(-Check) is happy.

What we additionally have to check: What happens when there's no internet connection available?

LineAgeOS 17 (Android 10) on my old Samsung S5neo does not allow to change that setting, it just reports that the application requested it (same as with original Android 6 firmware). But there I haven't experienced this bug (until now).

**Comment by SandraEickel** [comment]
I did two more reboots, with and without Autostart enabled and running CovPass-Check not earlier than two hours after reboot.
Results as expected, with Autostart correct initialization, without complaint about time.
For my Xiaomi, the only still outstanding test is a reboot in flight mode (with Autostart enabled), but that's more like a very special case anyway ...

Expecting to hear about Google Pixel and other affected phones ...

**Comment by wkornewald** [comment]
@SandraEickel Very nice. Now we have some confirmation for Xiaomi devices. Other device types will probably have to wait for one of the upcoming releases to get this fixed in a device-independent way.

**Comment by SandraEickel** [comment]
Reboot with flight mode was no problem (CovPass-Check did tell me that I was offline - after going back to normal operation, everything normalized within less than a minute).

**Comment by RobTho** [comment]
Still happens with 1.16.1 on a Pixel 4a

**Comment by Jo-Achim** [comment]
Today **25.01.2022 14:47** _CovPassCheck (v 1.16.1) (Android 11)_ on "Overview" shows again "Check date and time"; should be: **24.01.2022 01:48**). 
WiFi internet connection available.

Closing and restarting CovPassCheck brought no changes. 
Only restarting the phone solved this problem.

- - - -

_Steps I have taken following the "Check date and time" notice above:_ 
Manually "Update validation rules", result: "Last update ... 25.01.2022 14:49 ... Validation rules 25.01.2022 14:49" is displayed.
After 2x "<-" the above problem "Check date and time" is still displayed on the "Overview" page!

Even after restarting CheckCovPass, the message still exists. But now with "**24.01.2022 01:57**".
Scrolling down further into the field "Offline mode" is "Offline mode available with "List of certificate issuers: 25.01.2022 14:49 | Check rules: 25.01.2022 14:49. This was correct; see above.

After some minutes on "Overview": "Check date and time", should be: 24.01.2022 **02:03**.
After some minutes on "Overview": "Check date and time", should be: 24.01.2022 **02:13**; real time (phone): 15:12.

Closing CovPassCheck and restart: "Overview": "Check date and time": target: 24.01.2022 **02:17**); real time (phone): 15:14.

After restart the phone (_Samsung Galaxy Note 10, SM-N970F/DS / Android 11 (One UI 3.1)_): the problem "Check date and time" is solved; see above.

**Comment by Ein-Tim** [comment]
When will this issue be fixed?

**Comment by pucherd** [comment]
Hello @Ein-Tim, it is the same bug as #119. It is fixed and will be available in 1.18.0 release.

**Comment by Ein-Tim** [comment]
Thanks!

**Comment by RobTho** [comment]
still happen on Pixel 4a with new version 1.18.0

**Comment by jkrwdf** [comment]
With the new 1.18.0 arriving today, the message on my phone is gone. 

Although the absence of the message is no conclusive proof (this would need a couple of days and longer phases of a fully switched-off device), the mentioning of this issue in https://github.com/Digitaler-Impfnachweis/covpass-android/discussions/187 ("NTP-Timeserver Fix [Android]") gives me hope.

I refrain from closing due to the comment of @RobTho and wait for further instructions ;-)

**Comment by RobTho** [comment]
after **2**x reboot of the phone (after 1.18.0 was installed) , the issue is currently gone. Need to check in next days if the issue reappear

**Comment by jkrwdf** [comment]
Scheduled for closing on Monday, February 21, 2022 (at "High Noon")

**Comment by jkrwdf** [comment]
My Pixel 3 XL was switched off during the night. 
Rebootet this morning. 

Unfortunately, CovPassCheck 1.18.0 greets me as such: 

![image](https://user-images.githubusercontent.com/5329491/154629994-1a5c0e53-d36c-485f-a24e-87c7dc00a87a.png)

Remarkable is that the indicated difference is now exactly "24h 0m" (see clock at top left screen and indicated time, both are 7:26, just the date (today is 18.2., CovPassCheck says it should be 17.2.) differs.

**Comment by RobTho** [comment]
my Pixel 4a also show this when switched off during night.

**Comment by MikeMcC399** [comment]
I confirm that this issue is **not** reliably fixed.
This is a follow-on from my previous post https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issuecomment-1021368149.

CovPassCheck App 1.18.1 on Samsung Galaxy A52 5G SM-A526B/DS, Android 12. 

The correct date and time shown by the system of the mobile device is Feb 18, 2022 19:17.

CovPassCheck on the other hand was for some reason convinced that the correct date is Feb 4. This happened after switching the device off for one hour. After going to the info page of the app via the (?) symbol to check for the app version and returning to the main screen of the app the warning message regarding "Check date and time" had disappeared.

**Comment by timokoenig** [comment]
Since this is a very common problem after the system reboots, it might be an easy fix to include a grace period after the reboot where this message doesn't show up. Within that period the app could request multiple times the ntp server until it has a more reliable result

https://developer.android.com/reference/android/os/SystemClock#uptimeMillis()

**Comment by MikeMcC399** [comment]
@timokoenig 
Before displaying a warning, there should be a double-check in real-time with an ntp server (independently of whether the device has just been rebooted). If it is so obvious to a user that the warning is false, then it ought to be obvious to the app as well.

**Comment by Jo-Achim** [comment]
Report from today, 02/28/2022, CovPassCheck 1.18.0...
(Samsung Galaxy Note 10 / Android 12 (One UI 4.0))

**Check date and time**  {18:03:08}
Please correct the date and time of your smartphone.
Target: 02/18/2022 **_12:23_** p.m

Update performed manually...

**Update completed.**   {18:03:57}
List of certificate issuers: 02/28/2022 18:03
Check rules: 02/28/2022 18:03

**Check date and time**  {18:04:19}
Please correct the date and time of your smartphone.
Target: 02/18/2022 **_12:28_** p.m 

--------------

**Check date and time**  {18:40}
Please correct the date and time of your smartphone.
Target: 02/18/2022 **_01:04_** p.m 

Closing all apps.

Reopen CovPassCheck 1.18.0 {18:42}
**Check date and time**  
Please correct the date and time of your smartphone.
Target: 02/18/2022 **_01:06_** p.m 

Closing all apps and restart the phone. 

Reopen CovPassCheck 1.18.0 {18:45}
No more date/time error message.

**Comment by RobTho** [comment]
still happens with 1.19.0 on a Pixel 4a (when Phone was powered off over night)

**Comment by MikeMcC399** [comment]
The "Check date and time" function should be disabled if it cannot be made to work reliably.

Today, March 3, 2022, I am temporarily advised by CovPassCheck 1.19.0 that the "correct" date is 26 Feb 2022! (Same device as already reported in https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issuecomment-1021368149.) Security policy prevented taking a screenshot.

**Comment by Ein-Tim** [comment]
I support @MikeMcC399. If this feature can't be fixed, it should be deactivated until you found a way to get it to work reliable.

**Comment by MikeMcC399** [comment]
According to CovPassCheck (1.19.0) it is now 30 Jan 2022 09:14. The device has a current uptime of 166:56:11 i.e. about 6 days since last reboot. This is on a Samsung Galaxy A50 (SM-A505FN/DS) running Android 11. It is in fact 11 Mar 2022, not 30 Jan 2022. 🙁

**Comment by Ein-Tim** [comment]
@pucherd What is your plan here? Will you just leave the buggy implementation in the app? Will you deactivate the feature until a more robust solution is found?

**Comment by timokoenig** [comment]
I just had a minute to investigate this issue a little bit, here are some ideas:

* Why does the app use `1.de.pool.ntp.org` and not `pool.ntp.org`? `pool.ntp.org` already chooses the nearest servers around you which could in some cases be better than always using the German pool.
* The used library has many open issues which are unanswered for years - This doesn't seem to be actively maintained like they say
* I was able to query the ntp servers from my MacBook where I saw some strange behaviors. The tool requests 5 servers from the pool and sometimes I receive one or more failed requests that result in an offset between a couple hours and multiple days. On my MacBook this doesn't seem to be the issue because those failed requests get filtered out. I have the feeling that the library that is being used doesn't do that. This would result in exactly the issue mentioned in this ticket. I'm not gonna dig deeper into the library but this could be a starting point for you.
* Maybe try to use the Android version of the library that iOS CovPass uses (Kronos: https://github.com/lyft/Kronos-Android)

**Comment by pucherd** [comment]
Hello @Ein-Tim, sorry for the later response. The fix for this bug is planned to be release in 1.22 version.

@timokoenig Thank you for the hint - we also tend to use the same approach as iOS.

**Comment by MikeMcC399** [comment]
@pucherd 
> The fix for this bug is planned to be release in 1.22 version.

Thanks for confirming that there is a fix planned, even if it is several versions into the future.

I notice that you already suspected a bug in the [TrueTime for Android](https://github.com/instacart/truetime-android) library  when you made your [comment ](https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issuecomment-971530808) in this issue back in November 2021.

Apart from rebooting the device as a workaround, deleting the CovPassCheck app's data (Settings > Apps > `CovPass Check` > Storage > Clear data) also gives a chance to clear the condition (with the disadvantage that the onboarding sequence is also triggered).

**Comment by MikeMcC399** [comment]
@timokoenig 
> I was able to query the ntp servers from my MacBook where I saw some strange behaviors. The tool requests 5 servers from the pool and sometimes I receive one or more failed requests that result in an offset between a couple hours and multiple days. On my MacBook this doesn't seem to be the issue because those failed requests get filtered out. I have the feeling that the library that is being used doesn't do that. This would result in exactly the issue mentioned in this ticket. I'm not gonna dig deeper into the library but this could be a starting point for you.

The [Wiki Caveats - Gotchas](https://github.com/instacart/truetime-android/wiki/Caveats---Gotchas) for [TrueTime for Android](https://github.com/instacart/truetime-android) mentions that NTP runs on UDP and therefore the app needs to look for `InvalidNtpServerResponseException`  or use the "Rx-ified version of the library". "This means it is highly plausible that we get faulty data packets.".  

I may have easily have missed something reading the code for CovPassCheck, but I couldn't see it handling failed requests. The library used is https://github.com/Digitaler-Impfnachweis/covpass-android/blob/06b28d2a79f5cad73a89ed9cd88991bdf059db5e/dependencies.gradle#L213
which is not the Rx-ified version.

**Comment by timokoenig** [comment]
@MikeMcC399 thanks for sharing! The statement `These guards are extremely important to guarantee accurate time and cannot be avoided.` should've been placed a bit more prominent 😄

**Comment by abrock** [comment]
I also have this issue (CovPassCheck 1.19, moto g 5G plus, Android 11).
It is 22-03-24, 21:17 and it says "It should be: Feb 12, 2022 10:29 AM).

When is 1.22 expected to be released?
Does this affect the result shown for recently expired certificates?

**Comment by Ein-Tim** [comment]
1.22 is only in 3 versions. This could take until the end of April or even the middle of May, I assume.

**Comment by Jo-Achim** [comment]
Samsung Galaxy Note 10, SM-N970F/DS
Android 12 (One UI 4.0)
CovPassCheck 1.19.0

Since I restarted my smartphone a few days ago (about 1 week), CovPassCheck keeps showing me the message "Check date and time".

Attached is a corresponding screenshot from today with the request to heed the message "Check date and time" and the correct date and time in the "Offline mode" field at the same time.

![IMG_20220330_221113_1](https://user-images.githubusercontent.com/21124877/160925313-35bc7bb5-bf54-4dfd-aca0-040d59f8fdd6.jpg)

PS: By the way, after "Update validation rules" I see the same error message "Error 902" as on CovPass v1.21.0. See: [Error 902: at "Update entry rules" ("Einreiseregeln aktualisieren") ...](https://github.com/Digitaler-Impfnachweis/covpass-android/issues/200#issue-1186737845).

**Comment by MikeMcC399** [comment]
@Jo-Achim 
> PS: By the way, after "Update validation rules" I see the same error message "Error 902" as on CovPass v1.21.0. See: https://github.com/Digitaler-Impfnachweis/covpass-android/issues/200#issue-1186737845.

- I suggest that you post that information directly in #200.

**Comment by pucherd** [comment]
@Jo-Achim are you able to reproduce the 902 error now? Because I tried it locally and I am not able to reproduce it.

**Comment by MikeMcC399** [comment]
- The conversation about https://github.com/Digitaler-Impfnachweis/covpass-android/issues/200 is continued in the issue devoted to updating rules. Discussions about "Update rules of entry" (CovPass) and "Update rule set" (CovPassCheck) are off-topic in this issue about date/time.

**Comment by MikeMcC399** [comment]
@Jo-Achim 
- Regarding your [post ](https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issuecomment-1083598665)about the date/time warning in CovPassCheck 1.19: it was already confirmed by @RobTho, @abrock and myself that the issue still occurs in the 1.19 version.  @pucherd wrote in https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issuecomment-1065835449 that a fix is planned for the 1.22 release.

**Comment by Jo-Achim** [comment]
> @Jo-Achim are you able to reproduce the 902 error now? Because I tried it locally and I am not able to reproduce it.

Yes, until I have reinstalled CovPassCheck v1.19.0; see: https://github.com/Digitaler-Impfnachweis/covpass-android/issues/200#issuecomment-1084565739.

**Comment by Jo-Achim** [comment]
_Information:_
Today (**April 08, 2022 21:35**) after Android update (on "Samsung Galaxy Note 10, SM-N970F/DS") from "Android 12 (_One UI 4.0_) with Android security update: February 01, 2022" to " Android 12 (_One UI 4.1_) with Android security update: March 01, 2022” and phone restart:
> CovPassCheck 1.19.1

means
> Date and Time
> Destination: **April 05, 2022 21:06**.

**Comment by MikeMcC399** [comment]
@Jo-Achim 

I'm wondering why you posted about the problem occurring on CovPassCheck 1.19.1.

My understanding from reading https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issuecomment-1065835449 is that a fix isn't expected until 1.22 is released.

Am I missing something?

**Comment by Jo-Achim** [comment]
That's easy, @MikeMcC399.

On the one hand I wanted to underline the need to fix the bug soon and on the other hand I wanted to give more information. Whether this information is actually still needed by the developers, I'll leave it open.
(It may also be that I actually have little understanding of how long it takes to fix the problem (more than 5 month). If it had occurred in one of my applications, at least the priorities would have been different.)

**Comment by pucherd** [comment]
Hello @jkrwdf, 1.22 is released

**Comment by jkrwdf** [comment]
I confirm that in 1.22 on both of my devices, the message is not shown. 

_Unfortunately_, today with the previous version (1.19.x), the message was also not shown (this sometimes happened, the reported issue was not always reproducible), therefore in the sense of "Absence of indication is no indication of absence" this is no terminal proof of resolution.

In the positive expectation that the situation is resolved, I close the issue and will re-open when the issue comes back in the next few days for version 1.22 or higher for me or others reporting here in further posts.

**Comment by Jo-Achim** [comment]
Of course, jumping from v1.19.x to v1.22.0 can also shorten the time. ;-)

And the hint is correct
> ..., the reported issue was not always reproducible

But before the joy gets too big, here is a screenshot of version 1.22.0 from a Samsung Note 10 (Android 12 / 4.1) with the well-known error message:

![IMG_20220415_104047_1_2](https://user-images.githubusercontent.com/21124877/163550990-a9a6038c-88d1-4f15-92af-4f00ecf7b7bb.jpg)

(Update performed yesterday (04/14/2022) and tested again today after restarting the smartphone.)

However, I can add that the error message was gone after a few minutes.
Whether the problem is permanently solved is perhaps still open. In any case, it did not appear again after a restart.

**Comment by Jo-Achim** [comment]
Update from today...
The "Check date and time" bug is **not** fully fixed in version 1.22.0!

On my Samsung Note 10 (Android 12) the error started appearing after restarting the phone. Closing CovPassCheck (after approx. 2 minutes) and restarting the app no longer displayed the error.

The same problem with the Nokia 7.2 with Android 11 (it tends to crash). After starting CovPassCheck v1.22.0 the "Check date and time" error was displayed. A tap on "?" and back solved the problem.

The problem cannot be directly reproduced on either smartphone.

I think the solution to the problem so far is not as it should be!?

**Comment by pucherd** [comment]
Hello @Jo-Achim, Thank you for reporting it. We will have a look and let you know in which release this behaviour will be fixed.

**Comment by jkrwdf** [comment]
Reopening based on recent reports

**Comment by Jo-Achim** [comment]
Thanks @jkrwdf.

_Information:_
Today, at 04/26/2022, 12:42, the error "Check date and time" occurred again. Target: 04/24/2022 19:05.
Under 2G+ check the same message.

After tabbing on the question mark (?) and back to the main screen (<-), the error message was immediately cleared.

PS: Again the note: Internet connections are only allowed via WLAN.

PS2: I'm just wondering if I'm currently the only one who still notices the problem!?

(Samsung Galaxy Note 10, SM-N970F/DS; Android 12 (One UI 4.1) with Android security update: March 01, 2022; Google Play system update: March 01, 2022; CovPassCheck v 1.22.0.)

**Comment by Jo-Achim** [comment]
> @Jo-Achim
> 
> Did you have a stable WiFi connection **before** you opened CovPassCheck?

Yes, I had a stable WLAN before opening CovPassCheck - I was only a few meters next to my FRITZ!Box in the same room and in my home WLAN.

**Comment by MikeMcC399** [comment]
I have not seen this issue happening any more. Is it fixed or is anybody still seeing it with the latest CovPassCheck version 1.31.0?

**Comment by pucherd** [comment]
Hello @jkrwdf, if you agree we can close this issue.

**Comment by jkrwdf** [comment]
Agreed.
Closed.
Thanks to all.

**Comment by Ein-Tim** [comment]
@jkrwdf Did you reopen this issue by accident or do you experience the issue again?

**Comment by jkrwdf** [comment]
Unfortunately the latter. Sorry for the time difference between "Reopen" and this post with the new screenshot, it took some time to find out how hard it is with a mobile phone and that I have to start the big equipment, but here we go: 

![image](https://user-images.githubusercontent.com/5329491/193427181-64fdd068-2825-47c3-9390-c533c2557c10.png)

**Comment by ingowerren** [comment]
In 2021, the European Union introduced the digital COVID certificate as an EU-wide recognised proof of vaccination, recovery or negative testing for COVID-19. This regulation expired on 30 June 2023. On a transitional basis, the EU will continue to operate the technical systems to enable cross-border certificate checks until 31 December 2023. The transitional operation will end on that date.
In Germany, the possibility of issuing digital COVID certificates such as vaccination and recovery certificates will therefore be discontinued on 31 December 2023. The CovPass app and the CovPassCheck app will therefore also be discontinued. The CovPass app will be switched to wallet mode, where the stored certificates will be retained.

What does this mean for you?
All functionalities will continue to beavailable until 31 December 2023.

**From 1 January 2024:**

Support for the app will no longer be offered.

- The apps will no longer be available for (re)download in the app store.
- No new functions will be published.
- Further information can also be found in the CovPass app privacy policy.
